src/git/utils.rs

```diff
@@ -119,401 +23,117 @@ pub fn canonicalize_path(p: impl AsRef<Path>) -> Result<PathBuf> {
-                    ));
-                }
-            }
-            if e.code() != ErrorCode::NotFound {
-                return Err(e.into());
-            }
+    let repo = builder
+        .clone_with_submodules(git_clone_dir.path())
+        .context("Please check if the Git user / repository exists.")?;
+    let branch = get_branch_name_repo(&repo)?;
 
-            let path: &str = args.remote.as_ref();
-            let path = Path::new(path);
-            if !path.exists() || !path.is_dir() {
-                return Err(e.into());
-            }
+    // change from git repository into normal folder.
+    remove_history(git_clone_dir.path())?;
 
-            warn!("Template does not seem to be a git repository, using as a plain folder");
-            copy_dir_all(path, project_dir)?;
-            Ok("".to_string())
-        }
-    }
+    Ok((git_clone_dir, branch))
 }
 
-pub fn remove_history(project_dir: &Path, attempt: Option<u8>) -> Result<()> {
-    let git_dir = project_dir.join(".git");
-    if git_dir.exists() && git_dir.is_dir() {
-        if let Err(e) = remove_dir_all(git_dir) {
-            // see https://github.com/cargo-generate/cargo-generate/issues/375
-            if e.to_string().contains(
-                "The process cannot access the file because it is being used by another process.",
-            ) {
-                let attempt = attempt.unwrap_or(1);
-                if attempt == 5 {
-                    warn!("cargo-generate was not able to delete the git history after {} retries. Please delete the `.git` sub-folder manually", attempt);
-                    return Ok(());
-                }
-                let wait_for = Duration::from_secs(2_u64.pow(attempt.sub(1).into()));
-                warn!("Git history cleanup failed with a windows process blocking error. [Retry in {:?}]", wait_for);
-                sleep(wait_for);
-                remove_history(project_dir, Some(attempt.add(1)))?
-            }
-        }
-    }
-    Ok(())
-}
+// `cargo` itself has a problem with implementing git and it fallback to git cli
+// 1. https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
+// 2. https://github.com/rust-lang/cargo/blob/03e24bcf67696ba35d3aa2b0dd01f97bcfdf91a7/src/cargo/sources/git/utils.rs#L875
+// 3. https://github.com/rust-lang/cargo/issues/8172
+pub fn clone_git_using_cmd(
+    git: &str,
+    branch: Option<&str>,
+    identity: Option<&Path>,
+) -> anyhow::Result<(TempDir, String)> {
+    let temp_dir = tempfile::tempdir()?;
 
-pub fn init(project_dir: &Path, branch: &str, force: bool) -> Result<Repository> {
-    fn just_init(project_dir: &Path, branch: &str) -> Result<Repository> {
-        let mut opts = RepositoryInitOptions::new();
-        opts.bare(false);
-        opts.initial_head(branch);
-        Repository::init_opts(project_dir, &opts).context("Couldn't init new repository")
+    let mut cmd = Command::new("git");
```

Comment:
tbh it would be great if we would not shell out to git. I put already some arguments in the comment on the issue
https://github.com/cargo-generate/cargo-generate/issues/67#issuecomment-1032407568